### ISSUE TITLE
Expose enough of LibTxt to make custom text layout practical

### ISSUE THREAD

**Issue by suragch** [flutter/flutter#35994]
This is related to #224, #18656, and #35897, but for a specific purpose. It is also a followup from [this medium post](https://medium.com/@studymongolian/my-first-disappointment-with-flutter-5f6967ba78bf).

### Need

To be able to perform custom text layout beyond the currently supported ltr and rtl multiline paragraph layout.

### Example use cases

- Vertical layouts like [Mongolian and vertical CJK](https://www.w3.org/International/articles/vertical-text/)
- Writing text on non-linear paths (see also #16477)
- Filling a non-rectangular shape with text (examples [here](https://stackoverflow.com/questions/20476684/core-text-with-asymmetric-shape-on-ios) and [here](https://stackoverflow.com/questions/32547783/how-to-fill-triangle-in-swift-using-core-graphics))
- Doing custom text flow around objects (see Specifying Exclusion Paths [here](https://developer.apple.com/library/archive/documentation/StringsTextFonts/Conceptual/TextAndWebiPhoneOS/CustomTextProcessing/CustomTextProcessing.html))

### Current problem

The `ui.Paragraph` class is about all we have to work with. This class assumes you are doing multi-line, linear, horizontal layout within a rectangular shape. It seems probable that using a `ui.ParagraphBuilder` to measure every single glyph in a long string would be slow (but I haven't tested that yet). Many of the metrics related to the paragraph (like line heights and text) are not provided. As far as I know, there is no access to tools like the Minikin [LineBreaker](https://github.com/flutter/engine/blob/a4abfb2333afe16675db0709d18a6e918b0123da/third_party/txt/src/minikin/LineBreaker.cpp).

### Feature request

Expose just enough of the low level LibTxt API to make custom layout possible and practical. Based on [my experience](https://github.com/suragch/mongol-library) with custom text layout in Android, these are the features I think would be necessary to do custom text layout:

- [x] Get font metrics: ascent, descent, baseline, leading, line height
- [ ] Have access to a line/word breaker 
- [ ] Be able to measure text runs (including single characters/glyphs) efficiently
- [ ] Be able to measure substrings, similar to Android's `Paint.breakText` (see [example](https://stackoverflow.com/questions/45253574/does-paint-breaktext-include-the-character-at-the-maxwidth)). This is helpful for breaking a word that is longer than the max allowed length for the line. A character measured in the context of a word may have a different size than when it is measured individually. This is true for Mongolian, which like Arabic, has initial, medial, and final forms. Thus, when soft breaking a long word unnaturally across lines, it should render (and measure) the medial form at both sides of the break. Having something like Android's `Paint.breakText` helps with that. This one may be optional, though. I'd like to know if there is a better way to do it.
- [ ] Be able to paint short text runs efficiently

Note that this is not a feature request of any of the specific example use case I listed in the section above.

**Comment by suragch** [comment]
@GaryQian @Hixie This is a followup to my Medium post.

**Comment by GaryQian** [comment]
Hi there!
The following is what I proposed in a design doc, please let me know any feedback you have! This should allow you to obtain most of the metrics you need:

- **int getLineNumberForPosition(TextPosition position)**
    Returns the line corresponding to the position index.
- **int getNumLines()**
    Returns the total lines in the laid out text. If overflowed, return the max lines
- **FontMetrics getFontMetricsForPosition(TextPosition position)**
    Returns the FontMetrics corresponding to the position provided.
- **LineMetrics getLineMetrics(int line)**
    Returns the LineMetrics for the given line.
    Returns null if the line is negative or larger than the number of lines.

LineMetrics and FontMetrics are (roughly) defined as:
```
LineMetrics {
  Rect bounds;     // height, width, startX, endX, offset can
       // all be derived from this.
  String contents; // The text within this line. Could also be a byte buffer.
  double baseline; // The y position of the baseline for this line from the top of the paragraph.
  int lineNumber   // The line number where the first is 0

  // Mapping between text position ranges and the FontMetrics
  // associated with them.
  Map<Range, FontMetrics> fontMetrics;
}
```

```
FontMetrics {
  double Top; // distance to reserve above baseline
  double Ascent; // distance to reserve below baseline
  double Descent; // extent below baseline
  double Bottom; // extent below baseline
  double Leading; // distance to add between lines
  double AvgCharWidth; // average character width
  double MaxCharWidth; // maximum character width
  double XMin; // minimum x
  double XMax; // maximum x
  double XHeight; // height of lower-case 'x'
  double CapHeight; // height of an upper-case letter
  double UnderlineThickness; // underline thickness
  double UnderlinePosition; // underline position relative to baseline 
  double StrikeoutThickness; // strikeout thickness
  double StrikeoutPosition; // strikeout position relative to baseline
  String fontFamily;
  double fontSize;
}

```
The above proposed API will allow obtaining the following properties:

- Get the number of lines.
- Get the buffer slice corresponding to each text box.
- Get font and text metrics. The following are provided for free for each font:
  - Top: distance to reserve above baseline
  - Ascent: distance to reserve below baseline
  - Descent: extent below baseline
  - Bottom: extent below baseline
  - Leading: distance to add between lines
  - AvgCharWidth: average character width
  - MaxCharWidth: maximum character width
  - XMin: minimum x
  - XMax: maximum x
  - XHeight: of lower-case 'x'
  - CapHeight: height of an upper-case letter
  - UnderlineThickness: underline thickness
  - UnderlinePosition: underline position relative to baseline
  - StrikeoutThickness: strikeout thickness
  - StrikeoutPosition: strikeout position relative to baseline
- Get line number for index.
- Get the width/start+end of each line
- Get the height of each line
- Get the baseline positions for each line

The new methods of layout (path, vertical, etc) will require a more detailed look at to see if it is something we can/will/want to do.

**Comment by GaryQian** [comment]
Regarding the simple measurement capability, the way text measurement is implemented, full layout is actually performed on the text in order to provide the measurements. If you look at txt/src/minikin/Layout.cpp, there is a `doLayoutRunCached` which is called whenever any metrics from the text are needed. This means that any call that requires positions of glyphs to be know will end up doing a full layout anyways. Thus, a "measure" functionality is basically identical to layout and measure using post-layout measurement API (including the proposed above).

I dont remember the specific implementation of android's measure, but since Android is based off of minikin (albeit a different fork), I suspect the measure is only a convenience wrapper around laying out contents in a single line. The same functionality should be able to be achieved by building a paragraph, laying out, and returning the width.

**Comment by suragch** [comment]
@GaryQian  Your design doc proposal would be a huge help in doing custom layout!

In order to judge if this would be enough handle a wide range of custom layouts, let me go through the four examples I listed above. I'll say how I would naively try to do them. Could you give your feedback as to whether your proposal would handle these examples?

### Example 1: Vertical Mongolian

By the standard layout, four lines would look like this:

![mongolian_layout_1](https://user-images.githubusercontent.com/10300220/61083299-00f4cd80-a3f1-11e9-9dcc-e2e647201644.png)

For the Mongolian layout, the top line would be rotated 90 degrees clockwise and placed on the left. The second line would be rotated and placed to the right of the first line, and so on. The circled red characters would have to be rotated back (or handled as separate runs and never rotated at all). The correct layout would look like this (photo edited version):

<img width="212" alt="mongolian_layout_2" src="https://user-images.githubusercontent.com/10300220/61083462-72cd1700-a3f1-11e9-91e3-9a90a734e373.png">

I see two possible ways to do this: 

- Method 1: Layout the paragraph the usual way. Then use the line information to get the text in each line. Divide the text into two types of runs: Mongolian/English/Latin and CJK/emoji. Re-layout and print each run with the correct rotation, but keeping the original line length. Since CJK characters and emoji are approximately square, rotating them within the line should be acceptable. (Question: is the second layout necessary?)
- Method 2: Layout and paint the text word by word and emoji/CJK by emoji/CJK, the line length will be determined by the measured text length and and emoji height. It seems like with this method, having a word breaker (what Android calls a [BreakIterator](https://developer.android.com/reference/java/text/BreakIterator)) is missing.

### Example 2: Vertical CJK

This is the same image I used in [my Medium article](https://medium.com/@studymongolian/my-first-disappointment-with-flutter-5f6967ba78bf). Lines are made of characters stacked on top of each other from top to bottom and soft wrapping from right to left. Non CJK text would generally be handled within a line as in the Mongolian example above, but there would be various options for handling things like numbers. I've circled in red non CJK text that would need special handling.

![cjk_2](https://user-images.githubusercontent.com/10300220/61085213-50d59380-a3f5-11e9-9424-5d6868d2389e.png)

To layout this text I would layout/measure each CJK character one at a time. According to the sum of the glyph heights (or just use the font metrics directly), I would determine the length of the line. After I have the line length I would print each CJK character separately in the correct location. Most non-CJK characters could be handled as text runs.

Question: If I want to get the size of a single glyph (not the font metrics, but the actual glyph size) would I just layout a single CJK character and get the line metrics on it?

Is seems like with this method also, some sort of break finder is missing if I want to find possible line breaking locations for non-CJK text.

### Example 3: Filling a non-rectangular shape with text

Image from [here](https://www.raywenderlich.com/5960-text-kit-tutorial-getting-started)

<img width="173" alt="filled_shape" src="https://user-images.githubusercontent.com/10300220/61086200-d8bc9d00-a3f7-11e9-8c3f-8f1487365635.png">

I would get the line length requirements from the shape. Then I would measure the text word by word to compute the line wrapping locations. After that I would paint each line one at a time. 

Again, a tool to tell me where in the string potential like breaks are seems to be missing.

### Example 4: text flow around exclusion paths

Image from [here](https://developer.apple.com/library/archive/documentation/StringsTextFonts/Conceptual/TextAndWebiPhoneOS/CustomTextProcessing/CustomTextProcessing.html)

![textcontainer_2x](https://user-images.githubusercontent.com/10300220/61086525-b0816e00-a3f8-11e9-85b6-e194851a54aa.png)

I would handle this similarly to the previous example. Calculate how long each line segment is and then use a line breaker (if there were one) to find places to break the text into new line segments.

### Summary

I think your proposal would cover most situations (though having to use something called Line to get an individual glyph's measurements seems a little strange). However, a major thing that is missing is the LineBreaker. With the line breaker missing, I would be left with manually searching for spaces, but there are lots of other punctuation and characters where lines could be broken at. I once tried to implement my own line breaker, but it turned out to be so complicated that I gave up and just used the Android BreakIterator.

**Comment by GaryQian** [comment]
Thanks for all the input! I'll get back soon with a revised version of the API with these rounds of feedback integrated.

**Comment by GaryQian** [comment]
Goals for this initial round of changes is to expose what we have readily available. Some of the functionality such as potential break points is buried deeper within the stack, and will require a bit of work/time to cleanly extract. Many of the basic measurements can be rapidly consolidated and exposed though, and I would like to get that done first.

**Comment by frmatthew** [comment]
I too would appreciate the ability to layout a whole Paragraph and then render only a portion of it, as @suragch mentions.

For one app I'm working on, I'm creating hundreds of short Paragraphs for very small text snippets, which I need to measure and position all independently based on their text metrics. This step is currently one of the main bottlenecks in my app atm.

I have tested joining together all text, laying them out in one Paragraph (in my case on a single line), and then measuring them using getBoxesForRange; all of this ends up being faster (by a magnitude of ten) for measuring and positioning; but since I can't draw partial runs from the Paragraph, it's a moot point. If I could draw only portions of a Paragraph, I could minimize creating so many Paragraphs and speed things up quite a bit.

Thanks for considering the idea!

**Comment by GaryQian** [comment]
Thanks for the use case example!

I'll see what I can do!

**Comment by suragch** [comment]
I've started a Flutter package for Mongolian text layout. The source code is [here](https://github.com/suragch/mongol/blob/master/lib/mongol_paragraph.dart). 

- It currently handles line wrapping with line breaks allowed between space and non-space characters. I'll probably wait for the Minikin LineBreaker to handle more sophisticated line breaking.
- Each word is a run of text and is individually measured by laying it out as a Paragraph. I save these "Paragraphs" in a list. With the few tests I have done so far it seems ok, but I am expecting performance problems for longer passages.
- My implementation doesn't handle rotated emoji/CJK or even hard breaks yet.
- When the size constraints are smaller than a single word, there is no easy way to wrap the extra letters to the next line. I would need to layout every letter as a Paragraph in order to measure it. Since for Mongolian the glyph size depends on the surrounding characters, I would need to add a letter, insert a Zero Width Joiner, measure the partial word, and repeat this process until I have the correct length. The final text would also need to have Zero Width Joiners at both sides of the break. Inserting extra characters like this is not ideal.

**Comment by lvauvillier** [comment]
I have a new simple use case: enable the ability to set x and y coordinates of each characters.

Just like SVG:
```svg
<text x="0 15 27 41" y="0">Text</text>
```

Currently i have an inefficient workaround:
```dart
paintText({String text, TextStyle style, List<double> originX, Canvas canvas}) {
  for (int runeIndex = 0; runeIndex < text.runes.length; runeIndex++) {
    int charCode = text.runes.toList()[runeIndex];
    TextSpan span =
        TextSpan(style: style, text: String.fromCharCode(charCode));
    TextPainter tp = TextPainter(
        text: span,
        textAlign: TextAlign.left,
        textDirection: TextDirection.rtl);
    tp.layout();
    tp.paint(canvas, Offset(originX[runeIndex], 0));
  }
}
```
How can this use case fits in your API proposal ?

Is there a more efficient workaround ? The bottleneck is too many calls to layout().
I tried to make one pass to compute, layout and cache all textPainters, but this makes my UI laggy (10 000 characters to render).
As layout() is backed by a native function, i cannot put this code in an isolate.

Tips are welcomed :)

**Comment by GaryQian** [comment]
This is a WIP PR for some of the metrics I am looking at exposing in this initial round: https://github.com/flutter/engine/pull/10127

Hopefully this helps with what features to expect, and feel free to continue to provide feedback! It is nowhere close to done yet but I wanted to post it now to keep you updated.

**Comment by frmatthew** [comment]
Thanks for the start @GaryQian.

Is it necessary for `LineMetrics` to have `lineNumber`? If we get the metrics via `getLineMetrics(int line)`, seems redundant since we already which line we're indexing.

**Comment by GaryQian** [comment]
I think the (rough) idea was to expose something like `GetLineMetricsForIndex`, instead of a method only getting the line number for a text offset. Thus, the line number in the `LineMetrics` would be useful. Exactly what we expose is still in flux, but the core mini-refactor of the metrics will allow all this info to be accessible. `GetLineMetrics(int line)` would be accomplished through `GetLineMetrics()[line]` where GetLineMetrics returns a vector of all LineMetrics as this reveals more information than the former.

How does that sound to you?

Also, this is still just the LibTxt API, which may be slightly different than what we end up exposing to dart:ui. Primarily, the data here is a bit more "raw" and less packaged than what we may end up passing over to dart.

**Comment by frmatthew** [comment]
@GaryQian For the API and associated objects to be orthogonal, I'd think two calls would be clearer (eg, `GetLineForPosition()` followed by `GetLineMetrics(lineIndex)`). But I can't comment on the LibTxt API, so I'll defer to you... Thx.

**Comment by GaryQian** [comment]
There is an overhead for the switch from native to dart. For cpu light operations, I suspect it is more efficient to just simply provide all the data at once than to perform multiple dart to native to dart operations. A method returning only the line number would in practice be almost the same overhead as returning all information about the line, and making two separate calls would almost certainly be much slower in terms of performance.

Though this is not yet benchmarked, this generally seems to be the case. In any case, the idea is to somehow provide access to a vector of LineMetrics (or whatever the final name is), and a direct "get all the metrics" call seems to be the most straightforward way.

Another thing to consider is that the line number is actually used to perform layout, as I adapted LineMetrics from an existing internal class, so regardless, the line number is already there.

I hope this is helpful and may give some insight into how this should end up working.

**Comment by GaryQian** [comment]
Status update:

https://github.com/flutter/engine/pull/10127 has been adapted into a LibTxt-level-only change to properly track metrics. I will be creating a separate PR for the dart:ui level API.

**Comment by frmatthew** [comment]
Thanks @GaryQian, great to see the progress on the metrics portion of things. Has there been any thought also to @suragch's proposal to also be able to draw only portions of Paragraphs? I know this has some tricky aspects (what happens with line breaks, etc.), but I'd love to see how we can optimize drawing text for some of the uses described here, and drawing partial runs of texts would be key.

**Comment by GaryQian** [comment]
We have discussed it. I don't think a conclusion was reached though, I'll keep you updated though!

**Comment by frmatthew** [comment]
Here's an idea if you'll permit me. Haven't thought this through completely so feel free to critique.

If the bottleneck for our use cases above is the switch from native to dart for the creation of small Paragraphs, perhaps one solution to the expense of creating lots of Paragraphs is to have some batch process to create them at the native level, without the repeated context switching between dart and native. We wouldn't want to make a specialized API that can't be used in a variety of scenarios. But given that so much of mobile UI is short text runs, maybe there's justification for performance reasons.

For example:

```
    List<Paragraph> paragraphs = ui.BatchParagraphBuilder(ui.ParagraphStyle(/*...*/))
      ..addParagraph("Text block 1", textStyle: ui.TextStyle(/*...*/))
      ..addParagraph("Text block 2", textStyle: ui.TextStyle(/*...*/))
      // ...
      ..build(layoutConstraints: ui.ParagraphConstraints(width: double.infinity));
```

or even:

```
List<Paragraph> paragraphs = ui.BatchParagraphBuilder.build(ui.ParagraphStyle(/*...*/), List<...> runs);
```

Only the build() method would drop into native, combining the creation, building, and layout steps of the described text runs / paragraphs. Of course, this particular example limits the complexity of Paragraphs that could be created, and their general layout. But it seems that such text runs on mobile are very common, and could be used for all kinds of custom drawing at the dart:ui layer.

I guess I'm just thinking that perhaps specialized text layout layout described above (wrapping to curves, dynamic placement and drawing) could possibly be solved by batch creation of the Paragraphs rather than changing the drawing interface on the Paragraph.

**Comment by GaryQian** [comment]
The first part of the first part of this bug has been landed. https://github.com/flutter/engine/pull/10127 makes LibTxt track the metrics in a central `LineMetrics` object and provides access to it. It should also theoretically get a slight performance boost from reduced overhead, but this is unlikely to be significant/noticable. Next steps include exposing this through dart:ui and the framework.

Further features are still under discussion!

**Comment by suragch** [comment]
@GaryQian I'm looking forward to trying out the dart:ui version.

**Comment by GaryQian** [comment]
PR for dart:ui exposure: https://github.com/flutter/engine/pull/10670

This is still WIP.

**Comment by GaryQian** [comment]
Update on the exposing in dart:

After discussing the API, we realized that in the roadmap for the near future, we are looking to adopt Dart's new API for string and text handling (allowing for far better handling of things like emojis and extended grapheme clusters), which may very well not be based off of UTF-16 codepoints. Since this change would be breaking, we are currently going to hold off on adding UTF-16 codepoint based indexes to new API, as this API would eventually be broken in the transition.

What I can do is expose the metrics for the line that are agnostic to the text buffer index. These would be the per-line metrics and leave the index dependent API for a post-transition addition. The metrics that will be on hold would be:

- per-run raw metrics (the runs are indexed by UTF-16 codepoint indexes)
- start/end indexes

This would boil down to temporary hold on the `RunMetrics` class and the first few indexing variables in `LineMetrics`.

Meanwhile, we are looking for other ways to expose the RunMetrics and indexes, possibly through extension of TextBox or GetBoxesForRange, and if a suitable/superior API is found via those paths, we may be able to add it in sooner. Many of the metrics that are provided through `RunMetrics` can still be accessed in a slightly-processed form through the existing GetBoxesForRange call, which will return bonding boxes in a variety of formats for any index range.

**Comment by suragch** [comment]
I'm glad to hear that a better method for string indexing is being worked on. I've already run into [problems](https://stackoverflow.com/questions/54483177/handling-grapheme-clusters-in-dart) in the past with with emojis and extended grapheme clusters. Hopefully it will be easier to use than the clunky [Swift implementation](https://stackoverflow.com/questions/39676939/how-does-string-index-work-in-swift), but that may not be possible.

Your update sounds like you are going in in a wise direction. I'm not too concerned with getting the metrics right away since I am already able to compute my own metrics by building every word as a `Paragraph`, even though that feels like an expensive hack. Not having a `LineBreaker` is a bigger issue for my use case, but I realize you are discussing that and focusing on the metrics first.

**Comment by GaryQian** [comment]
Could you detail what an idea linebreaker api would look like in terms of dart-side use case code? We could consider something like a separate "text utilities" system that just performs relatively simple partial jobs such as linebreaking.

**Comment by suragch** [comment]
I'm fairly satisfied with the [`BreakIterator`](https://developer.android.com/reference/java/text/BreakIterator.html) in Android as I described it in [this post](https://stackoverflow.com/a/42219474/3681880). In Android I specifically use `BreakIterator.getLineInstance()` for finding line breaks. And `BreakIterator.getCharacterInstance()` is useful for dealing with grapheme clusters, but that probably relates to the same upcoming changes you were talking about earlier.

**Comment by w819314930** [comment]
About  int getNumLines() , When will it be updated?  In a widget_cell,the default is up to 3 lines of text.there's also a button to show or close all the 5 lines text, I need Implement the function dynamically.
Thank you for your work！

**Comment by GaryQian** [comment]
Update, metrics tracking had to be rolled back due to a breakage, but it is back in again. Working on getting the expose-api PR in again.

**Comment by GaryQian** [comment]
So the initial stage of resolving this should be well on its way. `TextPainter` should now expose measurements on a per-line basis through `computeLineMetrics`. Internally, we have additional information such as a per-run mapping of raw font metrics and raw start/end indexes prepared, but are withholding that in order to not break it in the near future.

We can move on to exploring the feasibility of some of the more advanced features requested here. I imagine a refactor can break out the portions of the code that handles breaking, and it may be able to be provided as utilities for more advanced developers.

Regarding painting short or partial lines of text, It should be technically possible, but we are weighing the API addition and performance (yet to be measured) compared to just creating numerous small paragraphs.

I will do my best to explore further in these areas to see what we can do!

**Comment by suragch** [comment]
@GaryQian  I was trying to find `computeLineMetrics` in [`TextPainter`](https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/painting/text_painter.dart) but I couldn't. Was I looking in the wrong spot?

**Comment by GaryQian** [comment]
Oh the PR is approved, just awaiting passing all presubmit tests. I will try to land the change in the morning 😊

**Comment by suragch** [comment]
@GaryQian I reviewed `LineMetrics` in an article [here](https://medium.com/@suragch/flutter-line-metrics-fd98ab180a64). Overall, I think its really good.

Also, [here is a recently published article detailing the process of custom text rendering](https://www.raywenderlich.com/4562681-flutter-text-rendering). It was written before `LineMetrics` came out, though.

Thank you for all of your hard work on this!

**Comment by longphanmn** [comment]
@GaryQian thanks for work. I'm waiting for this PR last couple months.

**Comment by suragch** [comment]
In [a comment above](https://github.com/flutter/flutter/issues/35994#issuecomment-520614495) I mentioned the Java `BreakIterator` being a good API. I still think the functionality there is good, but after learning more about Dart [Iterator](https://api.dartlang.org/stable/2.5.0/dart-core/Iterator-class.html) and [Iterable](https://api.dartlang.org/stable/2.5.0/dart-core/Iterable-class.html), I think it would be more Dart-like to follow that pattern.

**Comment by GaryQian** [comment]
Update on this issue, I am currently ooo at a conference, but I will continue this line of work when I return next week. Thanks for everyone's input and patience!

**Comment by suragch** [comment]
With the new [Dart 2.5 update](https://medium.com/dartlang/announcing-dart-2-5-super-charged-development-328822024970), will the `dart:ffi` foreign function interface for Dart-C interop allow referencing LibTxt directly without waiting for a Dart API?

**Comment by lwlizhe** [comment]
em,If we need to perform expensive calculations using textPainter，how to do it？
for example,[#30604](https://github.com/flutter/flutter/issues/30604)

**Comment by GaryQian** [comment]
It's been a while since the last update: Have worked through a few critical bugs for Samsung keyboards, as well as cranking through a few critical bugs involving font fallback and image cache memory usage. The features requested here are currently at the top of my "not critical" list of things to do.

**Comment by suragch** [comment]
@GaryQian Any updates on this?

**Comment by GaryQian** [comment]
Although not directly related, https://github.com/flutter/flutter/pull/44622 should expose the line boundaries information.

I have explored the realm of extracting out minikin breaker, athough have not made significant progress in the area. Will continue to work on this as time allows (recently, have been held up by blocking/critical samsung keyboard and iOS 13 bugs). Apologize for the delay!

**Comment by GaryQian** [comment]
We are also currently working with the Dart team on better grapheme cluster representations in dart, which will help solidify API in Flutter regarding more advanced characters and allow us to move forwards with some metrics/API regarding text buffer offests.

**Comment by Hixie** [comment]
This bug covers a wide variety of topics and isn't really directly tractable. It would be better to split this out into sub-bugs for each use case (e.g. vertical text, text-on-a-path, etc), because different use cases have different specific needs with different priorities.

I'm going to close this bug but please file bugs for each of the use cases you actually care about.

**Comment by suragch** [comment]
@Hixie 

I'm a little disappointed you closed this issue already. Here are my reasons:

- This is not a bug report, it's a feature request. We're asking for more freedom to handle text rendering ourselves, that is, more access to the low-level text capabilities of LibTxt. Note that this is not a feature request for any particular use case. I don't, for example, ever expect the Flutter team to support vertical text. What this is is a request to give developers the tools to handle text rendering problems themselves. These tools exist on Android and iOS, but not on Flutter.

- I specifically created this issue to generalize several use cases that have common requirements (with the possible exception of text-on-a-path). If I were to reopen a separate issue for each of them (vertical text, filling non-rectangular areas with text, text around exclusion paths), then I would be duplicating the same requirements. The reason is, as a developer the only way I know how to solve each of these problems to split a string into words, measure each word, and then draw it on the canvas in the correct location. This is quite difficult currently.

- We've been told that the way to show our support for a particular issue in the Flutter repo is to upvote that issue. Closing this issue and opening new smaller ones cancels all the support people have shown in the last six months. It means that this issue will no longer show up on the list of highly upvoted topics and thus will not be a priority.

For these reasons, I'd like to request that you reopen this issue. Even if you want to change the milestone to something more distant, having this issue open reminds you that we are still here.

Thank you for your consideration.

**Comment by Hixie** [comment]
The problem is that this issue is too vague for us to make useful decisions based on the thumbs up. As we tried to plan our work for 2020, we looked at this bug and couldn't figure out how to prioritise the work it asked for. Doing everything described here is a significant amount of effort, so we probably wouldn't do any of it if we had to do all of it or none of it. By splitting the requests into multiple narrowly-scoped bugs, we can prioritize the specific items of work that are most important.

For this reason, while we could reopen this issue, the net result would be that we would just ignore it. This isn't the best way for us to address your needs.

**Comment by suragch** [comment]
@Hixie 
Rather than dividing it by use case, would it help to divide it by feature request? There are four remaining unfinished tasks, but the highest priority one is having access to a line breaker.

**Comment by Hixie** [comment]
Use cases are preferred, because often people will list a solution and when we finally get a use case out of them it turns out their solution is massively more complicated than necessary. But if you prefer to do it by feature request, that works too, just make sure to include the use cases. :-)

**Comment by suragch** [comment]
I'm breaking the feature request in this issue into smaller pieces. The first one is #50171.

**Comment by suragch** [comment]
@frmatthew In your comments last year you mentioned that creating hundreds of small Paragraphs was a bottleneck in your app. Do you have an update on that? Is it still a performance issue? This relates to points 3 and 5 in this feature request. If you found some workaround then I won't focus on these issues. Otherwise I want to document them more specifically.

**Comment by GaryQian** [comment]
@suragch This is likely of interest to you:

We are currently evaluating and developing a process to switch to https://github.com/google/skia/tree/master/modules/skparagraph to replace LibTxt as the text shaper. This library is maintained by Skia and performs many of the same tasks as LibTxt does. We don't have a specific ETA yet, but are hoping for sooner rather than later, with a goal of transitioning with minimal (if any) disruption to existing apps while gaining performance and code cleanliness.

**Comment by suragch** [comment]
@GaryQian Yes, this is very much of interest to me! Where can I follow the progress of the switch?

**Comment by GaryQian** [comment]
https://github.com/flutter/flutter/issues/39420 Here is a bug that is tracking it, but as of now it has not been that active as a lot of the discussion has been in-person. I can try to make this process as transparent as possible though. A lot of the commits into the libtxt directory have been aimed at setting up/integrating skparagraph as well as converting over and adding tests/benchmarks to better validate the performance.

**Comment by suragch** [comment]
I added issue #51258 to address point 4 in this feature request. 

Regarding efficient measuring and painting (points 3 and 5), I haven't run into performance issues yet myself when creating many (i.e., hundreds of) small `Paragraph`s (as implemented in [this library](https://github.com/suragch/mongol)), so I'm not going to add bug reports for them until I have more concrete data.
